packages/intrepid2/unit-test/performance/StructuredIntegration/StructuredIntegrationPerformance.cpp

```diff
@@ -0,0 +1,953 @@
+  auto tensorCubatureWeights = cubature->allocateTensorCubatureWeights();
+  TensorPoints<PointScalar,ExecSpaceType> tensorCubaturePoints  = cubature->allocateTensorCubaturePoints();
+  
+  cubature->getTensorCubature(tensorCubaturePoints, tensorCubatureWeights);
+  
+  EOperator op = OPERATOR_GRAD;
+  BasisValues<Scalar,ExecSpaceType> gradientValues = basis->allocateBasisValues(tensorCubaturePoints, op);
+  basis->getValues(gradientValues, tensorCubaturePoints, op);
+  
+  // goal here is to do a weighted Poisson; i.e. (f grad u, grad v) on each cell
+    
+  int cellOffset = 0;
+  
+  auto jacobianAndCellMeasureTimer = Teuchos::TimeMonitor::getNewTimer("Jacobians");
+  auto fstIntegrateCall = Teuchos::TimeMonitor::getNewTimer("transform + integrate()");
+  
+  Data<PointScalar,ExecSpaceType> jacobian = geometry.allocateJacobianData(tensorCubaturePoints, 0, worksetSize);
+  Data<PointScalar,ExecSpaceType> jacobianDet = CellTools<ExecSpaceType>::allocateJacobianDet(jacobian);
+  Data<PointScalar,ExecSpaceType> jacobianInv = CellTools<ExecSpaceType>::allocateJacobianInv(jacobian);
+  TensorData<PointScalar,ExecSpaceType> cellMeasures = geometry.allocateCellMeasure(jacobianDet, tensorCubatureWeights);
+  
+  // lazily-evaluated transformed gradient values (temporary to allow integralData allocation)
+  auto transformedGradientValuesTemp = FunctionSpaceTools::getHGRADtransformGRAD(jacobianInv, gradientValues.vectorData());
+  auto integralData = IntegrationTools::allocateIntegralData(transformedGradientValuesTemp, cellMeasures, transformedGradientValuesTemp);
+  
+  const int numPoints = jacobian.getDataExtent(1); // data extent will be 1 for affine, numPoints for other cases
+  
+  // TODO: make the below determination accurate for diagonal/block-diagonal cases… (right now, will overcount)
+  const double flopsPerJacobianPerCell    = flopsPerJacobian(spaceDim, numPoints, numVertices);
+  const double flopsPerJacobianDetPerCell = flopsPerJacobianDet(spaceDim, numPoints);
+  const double flopsPerJacobianInvPerCell = flopsPerJacobianInverse(spaceDim, numPoints);
+  
+  transformIntegrateFlopCount = 0;
+  jacobianCellMeasureFlopCount  = numCells * flopsPerJacobianPerCell;    // jacobian itself
+  jacobianCellMeasureFlopCount += numCells * flopsPerJacobianInvPerCell; // inverse
+  jacobianCellMeasureFlopCount += numCells * flopsPerJacobianDetPerCell; // determinant
+  jacobianCellMeasureFlopCount += numCells * numPoints; // cell measure: (C,P) gets weighted with cubature weights of shape (P)
+  
+  initialSetupTimer->stop();
+  while (cellOffset < numCells)
+  {
+    int startCell         = cellOffset;
+    int numCellsInWorkset = (cellOffset + worksetSize - 1 < numCells) ? worksetSize : numCells - startCell;
+    int endCell           = numCellsInWorkset + startCell;
+    
+    jacobianAndCellMeasureTimer->start();
+    if (numCellsInWorkset != worksetSize)
```

Comment:
This is a bit ugly. we should find a way to resize, without reallocating, the data.

Reply:
OK, I think I've got something to address this.  I added a method to Data:
```
 /** \brief sets the notional extent in the specified dimension.  If needed, the underlying data container is resized.
     \param [in] d - the notional dimension in which the extent is to be changed
     \param [in] newExtent - the new extent
     \note Not supported for dimensions in which the variation type is BLOCK_PLUS_DIAGONAL.
     \note If the variation type is MODULAR, the existing modulus must evenly divide the new extent; the underlying data structure will not be resized in this case.
     */
    void setExtent(const ordinal_type &d, const ordinal_type &newExtent);
```

and in TensorData:
```
//! Sets the extent of the first component.  Only valid when separateFirstComponent() returns true.  The intended use case is when the 0 dimension in first component represents a cell index, and the container is resized to match a workset size that does not evenly divide the number of cells.
    void setFirstComponentExtentInDimension0(const ordinal_type &newExtent);
```

And here's the revised snippet from StructuredIntegrationPerformance:
```
    if (numCellsInWorkset != worksetSize)
    {
      const int CELL_DIM = 0; // first dimension corresponds to cell
      jacobian.setExtent(    CELL_DIM, numCellsInWorkset);
      jacobianDet.setExtent( CELL_DIM, numCellsInWorkset);
      jacobianInv.setExtent( CELL_DIM, numCellsInWorkset);
      integralData.setExtent(CELL_DIM, numCellsInWorkset);
      
      // cellMeasures is a TensorData object with separateFirstComponent_ = true; the below sets the cell dimension…
      cellMeasures.setFirstComponentExtentInDimension0(numCellsInWorkset);
    }
```

I've added a test case to the unit-test mode of StructuredIntegrationPerformance in which the worksetSize does not evenly divide the cell count.  This runs successfully; it's not a full test of the new code, but it at least exercises the new methods.

Reply:
Thanks! I think this is better. Do you think we could have a method in the geometry class that does all that? I.e. check if the numCellsInWorkset is different than worksetSize and if so it resizes the fields that need to be resized.  Not sure if this makes sense. Also it can happen in a later PR.